File: aria-backend/src/aria/connectors/bedrock.ts
import { BedrockRuntimeClient, InvokeModelCommand } from "@aws-sdk/client-bedrock-runtime";

import { ariaConfig, isConnectorLive } from "../config";
import { CrossServiceLog, Hypothesis, LogFinding, MetricPoint, Runbook } from "../types";

// ── Input / output types ──────────────────────────────────────────────────────

export interface BedrockRcaInput {
  // Alert snapshot
  service: string;
  summary: string;
  p99LatencyMs: number;
  errorRatePct: number;
  // Full log findings — includes stack traces, error kinds, trace IDs
  topErrors: LogFinding[];
  // Blast radius from Neo4j
  blastRadius: string[];
  // Historical runbooks from MongoDB Atlas
  runbooks: Runbook[];
  // Datadog enrichments
  tracesSummary: string;
  metricTrend?: string;
  spanSummary?: string[];
  errorRateSeries?: MetricPoint[];
  crossServiceLogs?: CrossServiceLog[];
}

interface BedrockRcaOutput {
  narrative: string;
  confidence: number;
  hypotheses: Hypothesis[];
  recommendedPlan: string[];
}

// ── Bedrock client singleton ──────────────────────────────────────────────────

let cachedClient: BedrockRuntimeClient | null = null;

function getClient(): BedrockRuntimeClient {
  if (!cachedClient) {
    cachedClient = new BedrockRuntimeClient({ region: ariaConfig.bedrock.region });
  }
  return cachedClient;
}

// ── Response parsing ──────────────────────────────────────────────────────────

function extractText(body: Uint8Array): string {
  const raw = JSON.parse(new TextDecoder().decode(body)) as {
    content?: Array<{ type?: string; text?: string }>;
  };
  return raw.content?.find((b) => b.type === "text")?.text ?? "";
}

function parseOutput(raw: string): BedrockRcaOutput | null {
  const jsonStart = raw.indexOf("{");
  const jsonEnd = raw.lastIndexOf("}");
  if (jsonStart < 0 || jsonEnd <= jsonStart) return null;

  try {
    const parsed = JSON.parse(raw.slice(jsonStart, jsonEnd + 1)) as BedrockRcaOutput;
    if (
      !parsed.narrative ||
      !Array.isArray(parsed.hypotheses) ||
      !Array.isArray(parsed.recommendedPlan)
    ) {
      return null;
    }
    return parsed;
  } catch {
    return null;
  }
}

// ── System prompt ─────────────────────────────────────────────────────────────

const RCA_SYSTEM = `You are ARIA, a senior on-call SRE performing root-cause analysis for a production incident.

Your job is to synthesize ALL available signals — error logs, stack traces, error trends, distributed trace data, historical runbooks, and blast radius — into a ranked list of hypotheses with concrete remediation steps.

Rules:
- Rank hypotheses by probability (highest first).
- Each hypothesis must reference specific signals from the evidence (log messages, error kinds, stack frames, trends).
- Remediation steps must be actionable and specific — not generic ("check logs").
- If a historical runbook closely matches the incident pattern, reference it explicitly.
- If error trend is "rising", treat the incident as actively escalating.
- Cross-service logs indicate blast radius propagation — factor them into impact assessment.

Respond ONLY with valid JSON (no markdown, no explanation):
{
  "narrative": "<2-3 sentences: what is happening and why, citing specific signals>",
  "confidence": <0.0–1.0>,
  "hypotheses": [
    {
      "title": "<specific hypothesis title — not generic>",
      "probability": <0.0–1.0>,
      "evidence": ["<specific signal from the data>", ...],
      "remediation": ["<concrete actionable step>", ...]
    }
  ],
  "recommendedPlan": ["<step 1>", "<step 2>", ...]
}`;

// ── Reasoner ──────────────────────────────────────────────────────────────────

export class BedrockReasoner {
  async synthesize(input: BedrockRcaInput): Promise<BedrockRcaOutput | null> {
    if (!isConnectorLive("bedrock")) return null;

    // Build a structured evidence package for the prompt
    const evidencePackage = {
      alertSnapshot: {
        service: input.service,
        summary: input.summary,
        p99LatencyMs: input.p99LatencyMs,
        errorRatePct: input.errorRatePct,
      },
      errorLogs: input.topErrors.map((e) => ({
        message: e.message,
        errorKind: e.errorKind ?? null,
        count: e.count ?? null,
        stackTrace: e.stackTrace ? e.stackTrace.split("\n").slice(0, 6).join("\n") : null,
        traceId: e.traceId ?? null,
      })),
      errorTrend: input.metricTrend
        ? {
            direction: input.metricTrend,
            seriesSummary:
              input.errorRateSeries && input.errorRateSeries.length > 0
                ? `${input.errorRateSeries.length} buckets, last value: ${input.errorRateSeries[input.errorRateSeries.length - 1]?.value ?? "?"}`
                : "no series data",
          }
        : null,
      traceCorrelation: {
        summary: input.tracesSummary,
        crossServiceErrors:
          input.crossServiceLogs?.slice(0, 8).map((l) => ({
            service: l.service,
            message: l.message,
          })) ?? [],
      },
      spanMetrics: input.spanSummary ?? [],
      blastRadius: input.blastRadius,
      matchedRunbooks: input.runbooks.map((r) => ({
        title: r.title,
        summary: r.summary,
        topSteps: r.steps.slice(0, 3),
        lastUsed: r.lastUsedAt,
      })),
    };

    const command = new InvokeModelCommand({
      modelId: ariaConfig.bedrock.modelId,
      contentType: "application/json",
      accept: "application/json",
      body: JSON.stringify({
        anthropic_version: "bedrock-2023-05-31",
        max_tokens: 1500,
        temperature: 0.1,
        system: RCA_SYSTEM,
        messages: [
          {
            role: "user",
            content: [{ type: "text", text: JSON.stringify(evidencePackage, null, 2) }],
          },
        ],
      }),
    });

    try {
      const response = await getClient().send(command);
      if (!response.body) return null;

      const text = extractText(response.body as Uint8Array);
      return parseOutput(text);
    } catch (error) {
      console.warn(
        "BedrockReasoner: synthesis failed —",
        error instanceof Error ? error.message : error,
      );
      return null;
    }
  }
}

File: aria-backend/aria/connectors/bedrock.py
"""Direct Bedrock client — boto3 with IAM credentials or bearer token (AgentCore) fallback."""
from __future__ import annotations

import json
import logging

from aria.config import config

logger = logging.getLogger(__name__)

_client = None


def _bedrock():
    global _client
    if _client is None:
        import boto3

        _client = boto3.client("bedrock-runtime", region_name=config.aws_region)
    return _client


def _invoke_rest(model_id: str, body: str) -> str:
    """Fallback: call Bedrock Runtime via REST with AgentCore bearer token."""
    import httpx

    url = f"https://bedrock-runtime.{config.aws_region}.amazonaws.com/model/{model_id}/invoke"
    try:
        resp = httpx.post(
            url,
            headers={
                "Authorization": f"Bearer {config.bedrock_api_key}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            },
            content=body,
            timeout=30,
        )
        resp.raise_for_status()
        payload = resp.json()
        return payload["content"][0]["text"]
    except Exception as exc:
        logger.error("Bedrock REST invoke failed: %s", exc)
        return ""


def _invoke(model_id: str, messages: list[dict], system: str, max_tokens: int = 1200, temperature: float = 0.1) -> str:
    body = json.dumps(
        {
            "anthropic_version": "bedrock-2023-05-31",
            "max_tokens": max_tokens,
            "temperature": temperature,
            "system": system,
            "messages": messages,
        }
    )
    # Try boto3 (standard IAM credentials)
    try:
        resp = _bedrock().invoke_model(
            modelId=model_id,
            contentType="application/json",
            accept="application/json",
            body=body,
        )
        payload = json.loads(resp["body"].read())
        return payload["content"][0]["text"]
    except Exception as exc:
        logger.warning("Bedrock boto3 invoke failed (%s), trying bearer token.", exc)

    # Fallback: bearer token (AgentCore)
    if config.bedrock_api_key:
        return _invoke_rest(model_id, body)

    logger.error("Bedrock unavailable: no valid credentials.")
    return ""


def synthesize_rca(context: dict) -> dict | None:
    """Call Bedrock to synthesize ranked hypotheses and remediation plan."""
    if not config.connector_live("bedrock"):
        return None

    system = (
        "You are ARIA, a production incident root-cause analyst. "
        "Respond ONLY with strict JSON: "
        '{"narrative": string, "confidence": float 0-1, '
        '"hypotheses": [{"title": str, "probability": float, "evidence": [str], "remediation": [str]}], '
        '"recommendedPlan": [str]}. '
        "Rank hypotheses by probability descending."
    )
    prompt = f"Incident context:\n{json.dumps(context, indent=2)}"
    raw = _invoke(config.bedrock_model_id, [{"role": "user", "content": prompt}], system)

    start, end = raw.find("{"), raw.rfind("}")
    if start == -1 or end == -1:
        return None
    try:
        return json.loads(raw[start : end + 1])
    except json.JSONDecodeError:
        return None


def copilot_chat(user_prompt: str, history: list[dict], system: str) -> str:
    """Single-turn Bedrock call for copilot chat responses."""
    messages = [*history[-8:], {"role": "user", "content": user_prompt}]
    return _invoke(config.copilot_model_id, messages, system, max_tokens=600, temperature=0.2)
